packages/grafana-data/src/types/scale.ts

```diff
@@ -0,0 +1,68 @@
+export interface ScaledValue {
+  percent?: number; // 0-1
+  threshold?: Threshold; // the selected step
+  color?: string; // Selected color (may be range based on threshold)
+}
+
+export type ScaleCalculator = (value: number) => ScaledValue;
+
+export interface Threshold {
+  value: number;
+  color: string;
+  state?: string; // Warning, Error, LowLow, Low, OK, High, HighHigh etc
+}
+
+export enum ScaleMode {
+  Absolute = 'absolute',
+  Relative = 'relative', // between 0 and 1 (based on min/max)
+  Scheme = 'scheme', // Pick from D3 scheme
+}
+
+export interface Scale {
+  mode: ScaleMode;
+  scheme?: ColorScheme; // D3 schema lookup
+
+  // Must be sorted by 'value', first value is always -Infinity
+  thresholds: Threshold[];
+}
```

Comment:
Been thinking if we really need this level of nesting (instead of having these props on the FieldConfig directly). It's probably a good idea. the concept of ScaleEditor becomes more self contained. But otherwise I cold not find where we passed only the scale object instead we always pass the full fieldConfig so it does not look 100% necessary. We will probably add more scale related props so it's probably a good idea.

An alternative would be to separate the scale & scheme options and have something like 

``` 
export interface Scale {
  thresholdMode: 'Absolute' | 'Percentage'; 
  colorMode: 'thresholds' | 'scheme' | 'thresholds-interpolate';
  schemeName: string;
  thresholds: Thresholds[]; 
```

Reply:
Or maybe:
```
export interface Thresholds {
  mode: 'Absolute' | 'Percentage'; 
  thresholds: Threshold[]
};

export interface FieldColor {
  mode: 'fixed' | 'thresholds' | 'thresholds-interpolate' | 'scheme' | 'scheme-discrete' | 'category'; 
  fixedColor?: string; // When fixed, simple color
  schemeName?: string;
  discreteSteps?: number;
  categories?: Record<string,string>; // Used for mapping strings to color
};
```
details on the rest of the FieldColor object are vague -- but this would allow us to extend it not locked into the concept of 'scale'.

It also means that things have a meaningful scale (likely mapping to state == Low OK, High) but allows us to color it based on something like a gradient

Reply:
The test for if nesting makes sense is if the root object on `FieldConfig` should all be changed at once.  For example -- changing the threshold mode without changing the threshold values does not make sense.  Likewise, colouring by threshold or a constant does not require changing the thresholds.

Reply:
ok, I agree about the nesting (just wanted to make sure we thought it through) :) 

I like separating thresholds and field color like that, make it a little clearer to me.
I fear the UI that exposes all this though :) But I imagine we would start with something like this

Color from: Thresholds
Threshold value: Absolute
Threshold list 

think it makes sense to have it in this order, because if you change color mode to scheme I image the thresholds editor will change a bit to gray out or remove the color options.

Sorry if I extended this PR a bit, but think getting these & terms right could be important as it's going to be hard to change later. 

One thing that will look a bit ugly with the Thresholds nesting is thisconfig.thresholds.thresholds, always looks bad when you two nestings with the same name, and a plural name for something not an array. Hm.. we always have my hacky proposal of using the first threshold to house the mode option. But feels too hacky.

Reply:
> Sorry if I extended this PR a bit, but think getting these model here right could be important as it's going to be hard to change later.

For sure -- if this was easy we would have already done it :)